## - Install the Required Packages -

In [1]:
%pip install sentence-transformers -q
%pip install langchain -q
%pip install transformers -q
%pip install pinecone-client -q
%pip install faiss-cpu -q
%pip install SpeechRecognition -q
%pip install pyttsx3 -q
%pip install -q openai
%pip install -q -U langchain-community
%pip install -q gTTS
%pip install -q langdetect
%pip install -q ipywebrtc
%pip install -q pvrecorder
%pip install -q sounddevice
%pip install pandas
%pip install langid -q
%pip install pyaudio -q


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


In [2]:
# set API key .env
import os
from dotenv import load_dotenv
load_dotenv()

OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
mapbox_api_key = os.getenv("MAPBOX_API_KEY")
google_search_api_key = os.getenv("GOOGLE_SEARCH_API_KEY")
google_search_cse_id = os.getenv("GOOGLE_SEARCH_CSE_ID")

## - Load and Prepare JSON Data -

In [3]:
import json
import pandas as pd

# Load JSON file 1 
with open(r"D:\Project\Rimal_AI_Lastdataset.json", 'r', encoding='utf-8') as f:
    data = json.load(f)

# If it's a list of objects, convert to DataFrame for easier handling
if isinstance(data, list):
    rimal_df = pd.DataFrame(data)
else:
    rimal_df = pd.json_normalize(data)

# Load JSON file 2 
with open(r"D:\Project\arabic_poems_Lastdataset.json", 'r', encoding='utf-8') as f:
    data = json.load(f)

# If it's a list of objects, convert to DataFrame for easier handling
if isinstance(data, list):
    poems_df = pd.DataFrame(data)
else:
    poems_df = pd.json_normalize(data)


In [4]:
rimal_df.head()

id      type                  name  \
0   1  landmark                Al-Ula   
1   2      city                  Neom   
2   3   culture       Bedouin Culture   
3   4  landmark               Diriyah   
4   5   culture  Saudi Coffee (Gahwa)   

                                         description  \
0  Al-Ula is an ancient city located in northwest...   
1  Neom is a planned city in northwestern Saudi A...   
2  Bedouins are nomadic groups living primarily i...   
3  Diriyah is the historic birthplace of the Saud...   
4  Saudi coffee is a traditional beverage symboli...   

                                          vision2030  \
0  Al-Ula is a centerpiece of Saudi Arabia's Visi...   
1  Neom is a flagship project of Saudi Arabia's V...   
2                                                NaN   
3  The Diriyah Gate project is a key cultural ini...   
4  Promoted as part of intangible cultural herita...   

                                   coordinates  \
0  {'latitude': 26.3631, 'longitude': 37.5525}   
1  {'latitude': 27.9893, 'longitude': 35.0094}   
2                                          NaN   
3    {'latitude': 24.737, 'longitude': 46.576}   
4                                          NaN   

                                             climate  \
0  {'summer': '40-45°C', 'winter': '15-25°C', 'be...   
1  {'description': 'Coastal desert climate with m...   
2                                                NaN   
3  {'summer': '38-45°C', 'winter': '14-25°C', 'be...   
4                                                NaN   

                                       notable_sites  \
0  [{'name': 'Mada'in Saleh', 'description': 'A U...   
1                                                NaN   
2                                                NaN   
3  [{'name': 'At-Turaif District', 'description':...   
4                                                NaN   

                                          activities  \
0  [{'type': 'cultural', 'name': 'Al-Ula Arts Fes...   
1                                                NaN   
2                                                NaN   
3  [{'type': 'cultural', 'name': 'Heritage Tours'...   
4                                                NaN   

                             sustainable_initiatives  ...  \
0  [{'project': 'Al-Ula Development Project', 'de...  ...   
1                                                NaN  ...   
2                                                NaN  ...   
3  [{'project': 'Diriyah Gate Development', 'desc...  ...   
4                                                NaN  ...   

                                            language  \
0                                                NaN   
1                                                NaN   
2  {'arabic': 'Primary language with desert-relat...   
3                                                NaN   
4                                                NaN   

                                            clothing  \
0                                                NaN   
1                                                NaN   
2  {'men': 'White thobe made of cotton or light w...   
3                                                NaN   
4                                                NaN   

                                             rituals  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  [Preparation with cardamom and saffron, Servin...   

                      values role_in_unification birthplace  era achievements  \
0                        NaN                 NaN        NaN  NaN          NaN   
1                        NaN                 NaN        NaN  NaN          NaN   
2                        NaN                 NaN        NaN  NaN          NaN   
3                        NaN                 NaN        NaN  NaN    

In [8]:
poems_df.head()

title                  poet    era           region category  \
0          لامية العرب               الشنفرى  جاهلي  الجزيرة العربية      فخر   
1           نهج البردة             أحمد شوقي   حديث              مصر      مدح   
2   إذا المرءُ لم يدنس  سمية بن خالد الفزاري  جاهلي  الجزيرة العربية      فخر   
3             قفا نبكِ            امرؤ القيس  جاهلي  الجزيرة العربية     رثاء   
4  لا تسقني ماء الحياة               المتنبي  عباسي           العراق     حكمة   

                                                text  \
0  أقيموا بني أمي، صدورَ مطيكم ... فإني إلى قومٍ ...   
1  ريمٌ على القاعِ بين البانِ والعلمِ ... أحل سفك...   
2  إذا المرءُ لم يدنسْ من اللؤمِ عرضه ... فكلُّ ر...   
3  قفا نبكِ من ذكرى حبيبٍ ومنزلِ ... بسِقطِ اللوى...   
4  لا تسقني ماءَ الحياةِ بذلةٍ ... بل فاسقني بالع...   

                                         translation  
0  O sons of my mother, guide your mounts' chests...  
1  A gazelle on the plain between the branches an...  
2  If a man’s honor is unstained by disgrace, the...  
3  Stop, let us weep for the memory of a beloved ...  
4  Do not give me the water of life in humiliatio...

In [9]:
poems_df.columns

Index(['title', 'poet', 'era', 'region', 'category', 'text', 'translation'], dtype='object')

## Create Embeddings and FAISS Vector Store 

1. Extract and combine the most relevant fields
2. (including media) Generate embeddings using Sentence Transformers
3. Create and save a FAISS vector store

### for the first file

In [4]:
import json
import pandas as pd
from sentence_transformers import SentenceTransformer
from langchain.vectorstores import FAISS
from langchain.embeddings import HuggingFaceEmbeddings

# --- Step 1: Combine Relevant Fields (including media) ---
def combine_text_fields(row):
    parts = []
    for field in ['type', 'name', 'description', 'vision2030']:
        val = row.get(field)
        if pd.notnull(val):
            parts.append(str(val))
    for list_field in ['notable_sites', 'activities', 'sustainable_initiatives', 'main_projects', 'tourist_destinations']:
        items = row.get(list_field)
        if isinstance(items, list):
            for item in items:
                if isinstance(item, dict):
                    name = item.get('name') or item.get('project_name') or item.get('project')
                    desc = item.get('description')
                    if name or desc:
                        parts.append(f"{list_field[:-1].replace('_',' ').title()}: {name}. {desc}")
    if isinstance(row.get('values_and_beliefs'), dict):
        values = row['values_and_beliefs'].get('values', [])
        if isinstance(values, list):
            parts.append(f"Values: {', '.join(values)}")
    if isinstance(row.get('rituals'), list):
        parts.append("Rituals: " + ", ".join(row['rituals']))
    if isinstance(row.get('values'), list):
        parts.append("Values: " + ", ".join(row['values']))
    media = row.get('media')
    if isinstance(media, dict):
        for media_type in ['images', 'videos', 'audio']:
            urls = media.get(media_type, [])
            if isinstance(urls, list) and urls:
                parts.append(f"{media_type.title()}: {', '.join(urls)}")
    return " | ".join(parts)

texts = rimal_df.apply(combine_text_fields, axis=1).tolist()
metadatas = rimal_df.to_dict(orient="records")

# --- Step 2: Generate Embeddings and Save as LangChain FAISS Vector Store ---
embedding_model = HuggingFaceEmbeddings(model_name="sentence-transformers/all-MiniLM-L6-v2")
vectordb = FAISS.from_texts(texts, embedding_model, metadatas=metadatas)
vectordb.save_local("rimalai_faiss")

# Save the DataFrame as metadata for mapping results (optional)
rimal_df.to_csv('rimalai_metadata.csv', index=False)

print("Embeddings generated and FAISS vector store saved in 'rimalai_faiss' directory (LangChain format).")



C:\Users\HUAWI\AppData\Local\Temp\ipykernel_22940\1280804079.py:43: LangChainDeprecationWarning: The class `HuggingFaceEmbeddings` was deprecated in LangChain 0.2.2 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-huggingface package and should be used instead. To use it run `pip install -U :class:`~langchain-huggingface` and import as `from :class:`~langchain_huggingface import HuggingFaceEmbeddings``.
  embedding_model = HuggingFaceEmbeddings(model_name="sentence-transformers/all-MiniLM-L6-v2")


Embeddings generated and FAISS vector store saved in 'rimalai_faiss' directory (LangChain format).


In [5]:
from langchain.vectorstores import FAISS
from langchain.embeddings import HuggingFaceEmbeddings

embedding_model = HuggingFaceEmbeddings(model_name="sentence-transformers/all-MiniLM-L6-v2")
vectordb = FAISS.load_local(
    folder_path="rimalai_faiss",
    embeddings=embedding_model,
    allow_dangerous_deserialization=True
)

### for the second file

In [6]:
import pandas as pd
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.vectorstores import FAISS
from langchain.docstore.document import Document

# Combine relevant fields (like title, poet, text, translation) into one field
poems_df['combined'] = poems_df['title'] + " " + poems_df['poet'] + " " + poems_df['text'] + " " + poems_df['translation']
relevant_data = poems_df[['combined']]

# Initialize the HuggingFace model
embedding_model = HuggingFaceEmbeddings(model_name="sentence-transformers/all-MiniLM-L6-v2")

# Wrap the combined text in Document objects
documents = [Document(page_content=text) for text in relevant_data['combined']]

# Create a FAISS vector store from the documents and the embedding model
vectordb = FAISS.from_documents(documents, embedding_model)

# Save the FAISS vector store to a directory (this will store the index and metadata)
vectordb.save_local("poem_faiss")



In [7]:
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.vectorstores import FAISS

# Initialize the HuggingFace model
embedding_model = HuggingFaceEmbeddings(model_name="sentence-transformers/all-MiniLM-L6-v2")

# Load the FAISS vector store from the saved location
vectordb = FAISS.load_local(
    folder_path="poem_faiss",  # Path where the FAISS index is saved
    embeddings=embedding_model,  # The embedding model used for generating embeddings
    allow_dangerous_deserialization=True  # Allow loading the FAISS index safely
)

# Set up audio

In [8]:
from pvrecorder import PvRecorder
import wave
import struct

def record_audio(output_file="user_question.wav", duration=5):
    recorder = PvRecorder(frame_length=512)
    recorder.start()
    frames = []
    
    # Record for 'duration' seconds
    for _ in range(int(duration * 16000 / 512)):  # 16000Hz / 512 samples-per-frame
        pcm_frames = recorder.read()
        # Convert 16-bit integer samples to bytes
        bytes_frames = [struct.pack('h', frame) for frame in pcm_frames]
        frames.extend(bytes_frames)
    
    recorder.stop()
    with wave.open(output_file, 'wb') as f:
        f.setnchannels(1)
        f.setsampwidth(2)  # 16-bit = 2 bytes
        f.setframerate(16000)
        f.writeframes(b''.join(frames))
    recorder.delete()

# Test recording
record_audio(duration=25)

# Agent and tools

##  poetry generation + image generation + text to speech + speech to text + QA with media

In [9]:
!pip install --upgrade openai

In [10]:
import os
import openai
import langid
import tempfile
import ast
import re
import os
import speech_recognition as sr
from gtts import gTTS
from IPython.display import Audio, display
from langchain.chat_models import ChatOpenAI
from langchain.vectorstores import FAISS
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.chains import RetrievalQA
from langchain.prompts import PromptTemplate
from langchain.tools import Tool
from langchain.agents import initialize_agent, AgentType



# === 1. Load saved FAISS ===
embedding_model = HuggingFaceEmbeddings(model_name="sentence-transformers/all-MiniLM-L6-v2")
vectordb = FAISS.load_local(
    folder_path="rimalai_faiss",
    embeddings=embedding_model,
    allow_dangerous_deserialization=True
)

# === 2. Initialize LLM ===
openai.api_key = os.getenv("OPENAI_API_KEY")
client = openai
llm = ChatOpenAI(model_name="gpt-4o", temperature=0.3)

# === 3. Define prompt templates ===
custom_prompt = PromptTemplate(
    input_variables=["context", "question"],
    template="""
    You are a helpful assistant. Answer the following question using the provided context.
    If the question is in Arabic, answer in Arabic. If in English, answer in English.

    Context:
    {context}

    Question: {question}
    Answer:
    """
)

# === 4. QA Chains setup ===
qa_chain = RetrievalQA.from_chain_type(
    llm=llm,
    chain_type="stuff",
    retriever=vectordb.as_retriever(),
    return_source_documents=True,
    chain_type_kwargs={"prompt": custom_prompt}
)

def query_with_media(query: str):
    response = qa_chain(query)
    answer = response["result"]
    media_results = []
    for doc in response["source_documents"]:
        media = doc.metadata.get("media", {})
        if isinstance(media, str):
            try:
                media = ast.literal_eval(media)
            except Exception:
                media = {}
        video_link = media.get("videos")[0] if media.get("videos") else None
        image_link = media.get("images")[0] if media.get("images") else None
        audio_link = media.get("audio")[0] if media.get("audio") else None
        media_results.append({
            "title": doc.metadata.get("name"),
            "image": image_link,
            "video": video_link,
            "audio": audio_link
        })
    return {"answer": answer, "media": media_results}

# === Tools Functions ===

# @tool 1: QA with Media
db_tool = Tool(
    name="SaudiRimalRetrievalQAWithMedia",
    func=query_with_media,
    description="Answer questions about Saudi landmarks, Vision 2030, and culture, returning answer and media links."
)

# @tool 2: Arabic Culture Poem + Image
ARABIC_CULTURE_KEYWORDS = [
    "arab", "saudi", "bedouin", "najd", "hijaz", "mecca", "medina", "riyadh", "alula", "al-ula",
    "najdi", "ghazal", "nabati", "jeddah", "diriyah", "desert", "camels", "dates", "gahwa", "coffee",
    "palms", "arabian", "oasis", "sand", "tribe", "poet", "poetry", "heritage", "tradition",
    "قصيدة", "شعر", "عربي", "سعودي", "بدوي", "نجد", "الحجاز", "مكة", "المدينة", "الرياض", "العلا",
    "جدة", "الدرعية", "صحراء", "جمل", "تمر", "قهوة", "نخيل", "قبيلة", "شاعر", "تراث", "تقاليد"
]

def is_arabic_culture(prompt):
    prompt_lower = prompt.lower()
    return any(kw in prompt_lower for kw in ARABIC_CULTURE_KEYWORDS)

def generate_arabic_poem(prompt):
    system_prompt = (
        "أنت شاعر عربي محترف. "
        "اكتب قصيدة عربية أصيلة حول الموضوع الذي يطلبه المستخدم، بأسلوب تقليدي (مثل النبطي أو الغزل أو الحكمة). "
        "بعد القصيدة، قدم ترجمة إنجليزية دقيقة لها تحت عنوان 'Translation:'."
    )
    user_prompt = (
        f"اكتب قصيدة عربية حول: {prompt}\n"
        "ثم ترجم القصيدة إلى الإنجليزية تحت عنوان 'Translation:'."
    )
    response = client.ChatCompletion.create(
        model="gpt-4o",
        messages=[
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt}
        ],
        max_tokens=400
    )
    return response.choices[0].message.content.strip()

def generate_image(prompt, n=1, size="1024x1024"):
    response = openai.Image.create(
        model="dall-e-3",
        prompt=prompt,
        n=n,
        size=size
    )
    return [img.url for img in response.data]

def arabic_culture_poem_and_image(prompt):
    if not is_arabic_culture(prompt):
        return {
            "message": (
                "⚠️ This tool is dedicated to Arabic culture and heritage. "
                "Please provide a prompt related to Arabic themes, history, or traditions."
            )
        }
    poem = generate_arabic_poem(prompt)
    images = generate_image(prompt)
    return {
        "poem": poem,
        "image_urls": images
    }

culture_tool = Tool(
    name="ArabicCulturePoemAndImage",
    func=arabic_culture_poem_and_image,
    description=(
        "Given a prompt related to Arabic culture, generates an original Arabic poem (with English translation) and a DALL-E image. "
        "If the prompt is not about Arabic culture, returns a polite message."
    )
)

# @tool 3: Text-to-Speech
def tts_tool_func(text: str) -> str:
    lang = langid.classify(text)[0]
    tts = gTTS(text, lang=lang if lang in ['en', 'ar'] else 'en')
    with tempfile.NamedTemporaryFile(delete=False, suffix=".mp3") as fp:
        tts.save(fp.name)
        return fp.name

tts_tool = Tool(
    name="GoogleTextToSpeech",
    func=tts_tool_func,
    description="Convert an answer in Arabic or English to speech (returns MP3 file path)."
)

# Voice Input Function (not a @tool)
def record_and_transcribe(duration=7, filename="user_question.wav"):
    r = sr.Recognizer()
    with sr.Microphone() as source:
        print("Listening...")
        audio = r.record(source, duration=duration)
        with open(filename, "wb") as f:
            f.write(audio.get_wav_data())
    with open(filename, "rb") as f:
        transcript = openai.Audio.transcribe("whisper-1,f")

    return transcript["text"]

# === 6. Agent Setup ===
tools = [db_tool, culture_tool, tts_tool]
agent = initialize_agent(
    tools=tools,
    llm=llm,
    agent=AgentType.OPENAI_FUNCTIONS,
    verbose=True
)

# === Voice Input Workflow ===
def record_audio(duration=7, filename="user_question.wav"):
    r = sr.Recognizer()
    with sr.Microphone() as source:
        print("Listening...")
        audio = r.record(source, duration=duration)
        with open(filename, "wb") as f:
            f.write(audio.get_wav_data())
    return filename

def extract_audio_path(agent_result):
    """Extract .mp3 file path from agent result string."""
    match = re.search(r'([/\w\-.]+\.mp3)', agent_result)
    if match:
        path = match.group(1)
        if path.startswith("sandbox:"):
            path = path.replace("sandbox:", "")
        if not os.path.exists(path) and os.path.exists(os.path.basename(path)):
            path = os.path.basename(path)
        return path
    return None

def voice_query_agent_workflow(agent, duration=7):
    audio_file = record_audio(duration=duration)
    print("Transcribing...")
    with open(audio_file, "rb") as f:
        transcript = openai.Audio.transcribe("whisper-1", f)
    user_text = transcript["text"]
    print("\nYou asked:", user_text)
    
    agent_prompt = f"Answer this question and return the spoken answer as an audio file: {user_text}"
    agent_result = agent.run(agent_prompt)
    
    audio_path = extract_audio_path(agent_result)
    if audio_path and os.path.exists(audio_path):
        print("\nPlaying answer:")
        display(Audio(audio_path))
    else:
        print("\nNo audio file found. Agent result:")
        print(agent_result)
    
    if isinstance(agent_result, str):
        print("\nAgent text result:", agent_result)
    else:
        print("\nAgent result:", agent_result)

def detect_language(text):
    lang = langid.classify(text)[0]
    return 'ar' if lang == 'ar' else 'en'

def voice_query_and_speak(agent, duration=7):
    audio_file = record_audio(duration=duration)
    print("Transcribing...")
    with open(audio_file, "rb") as f:
        transcript = openai.Audio.transcribe("whisper-1",f)
        user_text = transcript["text"]

    print("\nYou asked:", user_text)
    agent_result = agent.run(user_text)

    # Convert answer to speech
    lang = detect_language(user_text)
    tts = gTTS(agent_result, lang=lang)
    tts_file = "agent_reply.mp3"
    tts.save(tts_file)

    print("\nPlaying answer:")
    display(Audio(tts_file))

def text_query_and_reply(agent):
    user_text = input("Your question: ")
    print(f"\nYou asked: {user_text}")

    agent_result = agent.run(user_text)
    print("\nAgent (text answer):", agent_result)

    # Convert answer to speech
    lang = detect_language(user_text)
    tts = gTTS(agent_result, lang=lang)
    tts_file = "agent_reply.mp3"
    tts.save(tts_file)

    print("\nPlaying answer:")
    display(Audio(tts_file))

def unified_convo_agent():
    print("🎙️ Welcome to RimalAI! Ask about Saudi culture, landmarks, or poetry.")
    while True:
        mode = input("Type 'v' for voice, 't' for text, or 'q' to quit: ").strip().lower()
        if mode == 'q':
            print("Goodbye!")
            break
        elif mode == 'v':
            voice_query_and_speak(agent)
        elif mode == 't':
            text_query_and_reply(agent)
        else:
            print("Invalid option. Please type 'v', 't', or 'q'.")

C:\Users\HUAWI\AppData\Local\Temp\ipykernel_22940\1791125764.py:32: LangChainDeprecationWarning: The class `ChatOpenAI` was deprecated in LangChain 0.0.10 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-openai package and should be used instead. To use it run `pip install -U :class:`~langchain-openai` and import as `from :class:`~langchain_openai import ChatOpenAI``.
  llm = ChatOpenAI(model_name="gpt-4o", temperature=0.3)
C:\Users\HUAWI\AppData\Local\Temp\ipykernel_22940\1791125764.py:184: LangChainDeprecationWarning: LangChain agents will continue to be supported, but it is recommended for new use cases to be built with LangGraph. LangGraph offers a more flexible and full-featured framework for building agents, including support for tool-calling, persistence of state, and human-in-the-loop workflows. For details, refer to the `LangGraph documentation <https://langchain-ai.github.io/langgraph/>`_ as well as guides for `Migrating from AgentE

In [35]:
unified_convo_agent()

🎙️ Welcome to RimalAI! Ask about Saudi culture, landmarks, or poetry.

You asked: اعطني قصيدة عن القهوة


> Entering new AgentExecutor chain...

Invoking: `ArabicCulturePoemAndImage` with `قصيدة عن القهوة`


{'poem': "**القهوة العربية**\n\nيا فنجان القهوة في الصباح الباكر  \nريحك ينساب في روحي كالعنبر النادر  \n\nأنت للقلوب حديث يعشق السمر  \nويفيض منك الحب كما يفيض من نهرٍ جارف  \n\nفما بين عربك وقهوتك عهد زاهر  \nيا رمز الحضارة والعراقه علينا ان تذكر  \n\nأيتها الحبيبة ذات الطعم السكر  \nفيكِ روح السرور وراحة الخاطر  \n\nملاذ المشتاق والظمآن في محضر  \nتجلي الصدر وتفتح الفكر المبهر  \n\nكأنكِ للخواطر قيثارة شاعر  \nتنسج الأبيات وتبهر السامع الحاضر  \n\nنباهي بكِ في مجالس الوعي والسمر  \nفأنتِ للقلب حكاية لا تُهجر  \n\nتركتِ أثرًا في قلبي لا يُنكر  \nيا قهوة العرب يا نغمة الوتر  \n\n---\n\n**Translation:**\n\n**The Arabic Coffee**\n\nO coffee cup in the early morning  \nYour scent wafts in my soul like rare amber  \n\nYou are the conversation hearts cherish during evenings  \nAnd love

Goodbye!


## final modifications (poetry sreach + mapbox integration)

In [11]:
!pip install wikipedia-api

In [12]:
import os
MAPBOX_API_KEY = os.getenv("MAPBOX_API_KEY")


In [13]:
# mapbox integration

# ================== Imports ==================
import requests
import wikipediaapi
import ast
import json
import re
from langchain.vectorstores import FAISS
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.prompts import PromptTemplate
from langchain.chains import RetrievalQA
from langchain.chat_models import ChatOpenAI
from langchain.tools import tool
from langchain.agents import initialize_agent, AgentType
from langchain.agents import Tool as LangchainTool


# ================== 1. Load Vector Store ==================
embedding_model = HuggingFaceEmbeddings(model_name="sentence-transformers/all-MiniLM-L6-v2")
landmark_vectordb = FAISS.load_local(
    folder_path="rimalai_faiss",
    embeddings=embedding_model,
    allow_dangerous_deserialization=True
)


# ================== 2. Define RetrievalQA Chain ==================
llm = ChatOpenAI(model_name="gpt-4o", temperature=0.3)

custom_prompt = PromptTemplate(
    input_variables=["context", "question"],
    template="""
    You are a helpful assistant. Answer the following question using the provided context.
    If the question is in Arabic, answer in Arabic. If in English, answer in English.
    Context:
    {context}
    Question: {question}
    Answer:
    """
)

qa_chain = RetrievalQA.from_chain_type(
    llm=llm,
    chain_type="stuff",
    retriever=landmark_vectordb.as_retriever(),
    return_source_documents=True,
    chain_type_kwargs={"prompt": custom_prompt}
)

def query_landmark(query: str):
    response = qa_chain(query)
    answer = response["result"]
    media_results = []
    for doc in response["source_documents"]:
        media = doc.metadata.get("media", {})
        if isinstance(media, str):
            try:
                media = ast.literal_eval(media)
            except Exception:
                media = {}
        video_link = media.get("videos")[0] if media.get("videos") else None
        image_link = media.get("images")[0] if media.get("images") else None
        audio_link = media.get("audio")[0] if media.get("audio") else None
        media_results.append({
            "title": doc.metadata.get("name"),
            "image": image_link,
            "video": video_link,
            "audio": audio_link
        })
    return {"answer": answer, "media": media_results}


# ================== 3. Mapbox API ==================


def get_coordinates_from_mapbox(place_name):
    geocode_url = f"https://api.mapbox.com/geocoding/v5/mapbox.places/{place_name}.json"
    params = {
        "access_token": MAPBOX_API_KEY,
        "limit": 1,
        "country": "SA",  # Saudi Arabia
        "bbox": "34.5,16.0,55.7,32.2"
    }
    response = requests.get(geocode_url, params=params).json()
    features = response.get("features")
    if features:
        lon, lat = features[0]["center"]
        return lon, lat
    return None, None

def get_mapbox_image(place_name):
    lon, lat = get_coordinates_from_mapbox(place_name)
    if lon is None or lat is None:
        return f"https://api.mapbox.com/styles/v1/mapbox/streets-v11/static/0,0,1/600x600?access_token={MAPBOX_API_KEY}"
    url = f"https://api.mapbox.com/styles/v1/mapbox/streets-v11/static/{lon},{lat},5,0/600x600?access_token={MAPBOX_API_KEY}"
    return url


# ================== 4. Google Image Search ==================


def google_image_search(query):
    search_url = "https://www.googleapis.com/customsearch/v1"
    params = {
        "q": query,
        "cx": GOOGLE_CSE_ID,
        "key": GOOGLE_CSE_API_KEY,
        "searchType": "image",
        "num": 3
    }
    response = requests.get(search_url, params=params).json()
    return [item["link"] for item in response.get("items", [])]


# ================== 5. Wikipedia Search ==================
def search_wikipedia(query, lang="en"):
    wikipedia.set_lang(lang)
    try:
        page = wikipedia.page(query)
        return page.summary, page.images[:3]
    except Exception:
        return None, []


# ================== 6. Main Landmark Agent ==================
def landmark_agent(query):
    result = query_landmark(query)
    answer = result.get("answer")
    media = result.get("media")

    if answer and len(answer) > 20:  # Enough content
        return {
            "source": "vector_search",
            "title": query,
            "info": answer,
            "media": media,
            "map_url": get_mapbox_image(query)
        }
    else:
        # Fallback to Wikipedia + Google Images
        summary_en, wiki_images = search_wikipedia(query, lang="en")
        summary_ar, _ = search_wikipedia(query, lang="ar")
        google_images = google_image_search(query)
        return {
            "source": "wikipedia",
            "title": query,
            "info_en": summary_en,
            "info_ar": summary_ar,
            "images": google_images + wiki_images,
            "map_url": get_mapbox_image(query)
        }


# ================== 7. Tool Definition ==================
@tool
def landmark_search_tool(query: str) -> dict:
    """Search for information and media about Saudi landmarks. Accepts Arabic or English queries."""
    return landmark_agent(query)


# ================== 8. Initialize Router Agent ==================
llm = ChatOpenAI(model_name="gpt-4o", temperature=0)

landmark_tool = LangchainTool(
    name="landmark_search",
    func=landmark_search_tool,
    description="Useful for searching Saudi landmarks and tourist places. Accepts questions like 'Tell me about Al-Ula' or 'ما هي معالم الرياض؟'"
)

agent = initialize_agent(
    tools=[landmark_tool],
    llm=llm,
    agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION,
    verbose=True,
    handle_parsing_errors=True
)


# ================== 9. Extract Media & Map from Agent Response ==================
def extract_media_and_map(agent_response):
    try:
        # Find observation
        observation_start = agent_response.find("Observation: ")
        if observation_start == -1:
            # Fallback: look for final answer
            final_answer_start = agent_response.find("Final Answer: ")
            if final_answer_start == -1:
                return {"media": {"images": [], "videos": [], "audio": []}, "map_url": None}

            final_answer = agent_response[final_answer_start + len("Final Answer: "):]
            final_answer_end = final_answer.find("\n> Finished chain.")
            if final_answer_end != -1:
                final_answer = final_answer[:final_answer_end]

            image_urls = re.findall(r'\[.*?\]\((https://.*?\.png)\)', final_answer)
            map_urls = re.findall(r'\[.*?\]\((https://.*?mapbox.*?)\)', final_answer)

            return {
                "media": {
                    "images": image_urls,
                    "videos": [],
                    "audio": []
                },
                "map_url": map_urls[0] if map_urls else None
            }

        observation_text = agent_response[observation_start + len("Observation: "):]
        observation_end = observation_text.find("\nThought:")
        if observation_end != -1:
            observation_text = observation_text[:observation_end]

        observation = json.loads(observation_text)

        # Extract media
        media = observation.get('media', [])
        extracted_media = {"images": [], "videos": [], "audio": []}

        for item in media:
            if item.get('image'):
                extracted_media["images"].append(item['image'])
            if item.get('video'):
                extracted_media["videos"].append(item['video'])
            if item.get('audio'):
                extracted_media["audio"].append(item['audio'])

        map_url = observation.get('map_url')

        return {
            "media": extracted_media,
            "map_url": map_url
        }

    except Exception as e:
        print(f"Error extracting media and map: {str(e)}")
        return {"media": {"images": [], "videos": [], "audio": []}, "map_url": None}



In [14]:
# ================== 10. Example Usage ==================
agent_response = agent.run('Tell me about AlUla and show me some images and a map.')
print(agent_response)
media_and_map = extract_media_and_map(agent_response)

C:\Users\HUAWI\AppData\Local\Temp\ipykernel_22940\884815905.py:2: LangChainDeprecationWarning: The method `Chain.run` was deprecated in langchain 0.1.0 and will be removed in 1.0. Use :meth:`~invoke` instead.
  agent_response = agent.run('Tell me about AlUla and show me some images and a map.')




> Entering new AgentExecutor chain...
To answer the question about AlUla, I will first gather information about it using the landmark_search tool. This will help me provide a detailed description. However, I cannot directly show images or maps, but I can guide you on where to find them.

Action: landmark_search
Action Input: "Tell me about AlUla"

C:\Users\HUAWI\AppData\Local\Temp\ipykernel_22940\1763020866.py:52: LangChainDeprecationWarning: The method `Chain.__call__` was deprecated in langchain 0.1.0 and will be removed in 1.0. Use :meth:`~invoke` instead.
  response = qa_chain(query)



Observation: {'source': 'vector_search', 'title': 'Tell me about AlUla', 'info': "AlUla is a region in Saudi Arabia known for its rich history, stunning landscapes, and cultural significance. It is home to the Sharaan Nature Reserve, a protected sanctuary dedicated to preserving Arabian leopards and desert biodiversity. The reserve is developed as a global model for sustainable luxury eco-tourism, merging heritage conservation with environmental innovation. Notable sites within AlUla include the Sharaan Eco-Resort, a luxury resort designed by Jean Nouvel and carved into the natural rock, and the Arabian Leopard Conservation Zone, which is a secure area for the reintroduction and protection of endangered leopards.\n\nVisitors to AlUla can engage in various activities such as leopard spotting through guided eco-safaris, experiencing the Desert Sky Experience with astronomy tours under AlUla's clear night skies, and participating in Rock Trail Adventures with scenic treks through dramati

In [15]:
# poem search

from langchain.prompts import PromptTemplate
from langchain.chains import RetrievalQA
from langchain.chat_models import ChatOpenAI
from langchain.vectorstores import FAISS
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.tools import tool

# === 1. Load Vector Store ===
ppoetry_vectordb = FAISS.load_local(
    folder_path="poem_faiss",
    embeddings=embedding_model,  
    allow_dangerous_deserialization=True
)

# === 2. Load LLM ===
llm = ChatOpenAI(model_name="gpt-4o", temperature=0)

# === 3. Custom Prompt ===
custom_prompt = PromptTemplate(
    input_variables=["context", "question"],
    template="""
    You are a helpful poetry assistant.
    Use the following poems context to answer the question.
    If the question is in Arabic, answer in Arabic. If in English, answer in English.
    
    Context:
    {context}
    Question: {question}
    Answer:
    """
)

# === 4. Create RetrievalQA Chain ===
qa_chain = RetrievalQA.from_chain_type(
    llm=llm,
    chain_type="stuff",
    retriever=ppoetry_vectordb.as_retriever(),
    return_source_documents=True,
    chain_type_kwargs={"prompt": custom_prompt}
)

# === 5. Poetry Search Function ===
def poetry_search_agent(query: str):
    response = qa_chain(query)
    answer = response["result"]
    
    matched_poems = []
    for doc in response["source_documents"]:
        matched_poems.append({
            "title": doc.metadata.get("title"),
            "poet": doc.metadata.get("poet"),
            "category": doc.metadata.get("category"),
            "text": doc.page_content,
            "translation": doc.metadata.get("translation")
        })
    
    return {
        "answer": answer,
        "matched_poems": matched_poems
    }

# === 6. LangChain Tool Wrapper ===
@tool
def search_poetry_tool(query: str) -> dict:
    """Search for relevant Arabic poems based on query. Returns matching poems."""
    return poetry_search_agent(query)

# === 7. Example Usage ===
response = search_poetry_tool("أريد قصيدة عن الحب")
print(response["answer"])


C:\Users\HUAWI\AppData\Local\Temp\ipykernel_22940\254042786.py:71: LangChainDeprecationWarning: The method `BaseTool.__call__` was deprecated in langchain-core 0.1.47 and will be removed in 1.0. Use :meth:`~invoke` instead.
  response = search_poetry_tool("أريد قصيدة عن الحب")


بالطبع، إليك قصيدة عن الحب:

يا حبيبي في هواك القلبُ ذابْ  
مثل شمسٍ في غروبٍ أو شهابْ  

كلما ناديتُ باسمك في المساءْ  
أشرقَ الفجرُ على قلبي وضَاءْ  

أنتَ نبضُ الروحِ في دنيا الأماني  
أنتَ حلمُ العمرِ في ليلِ الزمانِ  

يا حبيبي، فيك أحلامي تطيبْ  
وبك الأشواقُ في قلبي تذيبْ  

فلتكنْ دوماً بقربي يا حبيبي  
ولتكنْ في كلِّ دربٍ لي نصيبي  


In [16]:
from langchain.prompts import PromptTemplate
from langchain.chains import RetrievalQA
from langchain.chat_models import ChatOpenAI
from langchain.vectorstores import FAISS
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.tools import tool

# === 1. Load Vector Store ===
ppoetry_vectordb = FAISS.load_local(
    folder_path="poem_faiss",
    embeddings=embedding_model,  
    allow_dangerous_deserialization=True
)



# === 2. Poetry Search Function (returns poems directly) ===
def poetry_search_agent(query: str):
    docs = ppoetry_vectordb.similarity_search(query, k=3)  # يرجع أفضل 3 قصائد (تقدر تغير العدد)
    
    matched_poems = []
    for doc in docs:
        matched_poems.append({
            "title": doc.metadata.get("title"),
            "poet": doc.metadata.get("poet"),
            "category": doc.metadata.get("category"),
            "text": doc.page_content,
            "translation": doc.metadata.get("translation")
        })
    
    return matched_poems

# === 3. LangChain Tool Wrapper ===
@tool
def search_poetry_tool(query: str) -> list:
    """Search for relevant poems based on query. Returns matching poems."""
    return poetry_search_agent(query)

# === 4. Example Usage ===
response = search_poetry_tool("أريد قصيدة عن الحب")
for poem in response:
    print(f"النص:\n{poem['text']}\n")


النص:
قفا نبكِ امرؤ القيس قفا نبكِ من ذكرى حبيبٍ ومنزلِ ... بسِقطِ اللوى بين الدخولِ فحوملِ Stop, let us weep for the memory of a beloved and a dwelling, at the edge of the sandy desert between Dukhool and Hawmal.

النص:
يا سائلي عنْ مودّتي ابن الفارض يا سائلي عنْ مودّتي ... إنّ مودّتيَ العَجبْ O you who ask about my love, indeed, my love is wondrous.

النص:
قفَا نبكِ من ذكرى حبيبٍ ومنزلِ امرؤ القيس قفا نبكِ من ذكرى حبيبٍ ومنزلِ ... بسِقطِ اللوى بين الدخولِ فحوملِ Stop, let us weep over the memory of a beloved and a dwelling at the remains of Al-Liwa between Dukhul and Hawmal.



# Intent classification based on user query (with all tools and agents)

In [17]:
import os
import openai
import langid
import tempfile
import ast
import re
import os
import speech_recognition as sr
from gtts import gTTS
from IPython.display import Audio, display
from langchain.chat_models import ChatOpenAI
from langchain.vectorstores import FAISS
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.chains import RetrievalQA
from langchain.prompts import PromptTemplate
from langchain.tools import Tool
from langchain.agents import initialize_agent, AgentType

# intent classification based on user query
def classify_user_intent(user_query: str) -> str:
    """
    Uses GPT-4o to classify user intent: text, poem, image, voice.
    """
    system_msg = (
        "You are an AI assistant that classifies user questions into one of the following intents: "
        "'text', 'poem', 'image', or 'voice'. "
        "Return ONLY one word without explanation."
    )
    user_msg = f"User asked: {user_query}\nWhat is the intent?"
    response = client.ChatCompletion.create(
        model="gpt-4o",
        messages=[
            {"role": "system", "content": system_msg},
            {"role": "user", "content": user_msg}
        ],
        max_tokens=1
    )
    return response.choices[0].message.content.strip().lower()


# === 1. Load saved FAISS ===
embedding_model = HuggingFaceEmbeddings(model_name="sentence-transformers/all-MiniLM-L6-v2")
vectordb = FAISS.load_local(
    folder_path="rimalai_faiss",
    embeddings=embedding_model,
    allow_dangerous_deserialization=True
)

# === 2. Initialize LLM ===
openai.api_key = os.getenv("OPENAI_API_KEY")
client = openai
llm = ChatOpenAI(model_name="gpt-4o", temperature=0.3)

# === 3. Define prompt templates ===
custom_prompt = PromptTemplate(
    input_variables=["context", "question"],
    template="""
    You are a helpful assistant. Answer the following question using the provided context.
    If the question is in Arabic, answer in Arabic. If in English, answer in English.

    Context:
    {context}

    Question: {question}
    Answer:
    """
)

# === 4. QA Chains setup ===
qa_chain = RetrievalQA.from_chain_type(
    llm=llm,
    chain_type="stuff",
    retriever=vectordb.as_retriever(),
    return_source_documents=True,
    chain_type_kwargs={"prompt": custom_prompt}
)

def query_with_media(query: str):
    response = qa_chain(query)
    answer = response["result"]
    media_results = []
    for doc in response["source_documents"]:
        media = doc.metadata.get("media", {})
        if isinstance(media, str):
            try:
                media = ast.literal_eval(media)
            except Exception:
                media = {}
        video_link = media.get("videos")[0] if media.get("videos") else None
        image_link = media.get("images")[0] if media.get("images") else None
        audio_link = media.get("audio")[0] if media.get("audio") else None
        media_results.append({
            "title": doc.metadata.get("name"),
            "image": image_link,
            "video": video_link,
            "audio": audio_link
        })
    return {"answer": answer, "media": media_results}

# === Tools Functions ===

# @tool 1: QA with Media
db_tool = Tool(
    name="SaudiRimalRetrievalQAWithMedia",
    func=query_with_media,
    description="Answer questions about Saudi landmarks, Vision 2030, and culture, returning answer and media links."
)

# @tool 2: Arabic Culture Poem + Image
ARABIC_CULTURE_KEYWORDS = [
    "arab", "saudi", "bedouin", "najd", "hijaz", "mecca", "medina", "riyadh", "alula", "al-ula",
    "najdi", "ghazal", "nabati", "jeddah", "diriyah", "desert", "camels", "dates", "gahwa", "coffee",
    "palms", "arabian", "oasis", "sand", "tribe", "poet", "poetry", "heritage", "tradition",
    "قصيدة", "شعر", "عربي", "سعودي", "بدوي", "نجد", "الحجاز", "مكة", "المدينة", "الرياض", "العلا",
    "جدة", "الدرعية", "صحراء", "جمل", "تمر", "قهوة", "نخيل", "قبيلة", "شاعر", "تراث", "تقاليد"
]

def is_arabic_culture(prompt):
    prompt_lower = prompt.lower()
    return any(kw in prompt_lower for kw in ARABIC_CULTURE_KEYWORDS)

def generate_arabic_poem(prompt):
    system_prompt = (
        "أنت شاعر عربي محترف. "
        "اكتب قصيدة عربية أصيلة حول الموضوع الذي يطلبه المستخدم، بأسلوب تقليدي (مثل النبطي أو الغزل أو الحكمة). "
        "بعد القصيدة، قدم ترجمة إنجليزية دقيقة لها تحت عنوان 'Translation:'."
    )
    user_prompt = (
        f"اكتب قصيدة عربية حول: {prompt}\n"
        "ثم ترجم القصيدة إلى الإنجليزية تحت عنوان 'Translation:'."
    )
    response = client.ChatCompletion.create(
        model="gpt-4o",
        messages=[
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt}
        ],
        max_tokens=400
    )
    return response.choices[0].message.content.strip()

def generate_image(prompt, n=1, size="1024x1024"):
    response = openai.Image.create(
        model="dall-e-3",
        prompt=prompt,
        n=n,
        size=size
    )
    return [img["url"] for img in response["data"]]

def arabic_culture_poem_and_image(prompt):
    if not is_arabic_culture(prompt):
        return {
            "message": (
                "⚠️ This tool is dedicated to Arabic culture and heritage. "
                "Please provide a prompt related to Arabic themes, history, or traditions."
            )
        }
    poem = generate_arabic_poem(prompt)
    images = generate_image(prompt)
    return {
        "poem": poem,
        "image_urls": images
    }

culture_tool = Tool(
    name="ArabicCulturePoemAndImage",
    func=arabic_culture_poem_and_image,
    description=(
        "Given a prompt related to Arabic culture, generates an original Arabic poem (with English translation) and a DALL-E image. "
        "If the prompt is not about Arabic culture, returns a polite message."
    )
)

# @tool 3: Text-to-Speech
def tts_tool_func(text: str) -> str:
    lang = langid.classify(text)[0]
    tts = gTTS(text, lang=lang if lang in ['en', 'ar'] else 'en')
    with tempfile.NamedTemporaryFile(delete=False, suffix=".mp3") as fp:
        tts.save(fp.name)
        return fp.name

tts_tool = Tool(
    name="GoogleTextToSpeech",
    func=tts_tool_func,
    description="Convert an answer in Arabic or English to speech (returns MP3 file path)."
)

# Voice Input Function (not a @tool)
def record_and_transcribe(duration=7, filename="user_question.wav"):
    r = sr.Recognizer()
    with sr.Microphone() as source:
        print("Listening...")
        audio = r.record(source, duration=duration)
        with open(filename, "wb") as f:
            f.write(audio.get_wav_data())
    with open(filename, "rb") as f:
             transcript = openai.Audio.transcribe("whisper-1", f)
    user_text = transcript['text']
    return transcript["text"]

# === 6. Agent Setup (Extended with New Tools) ===

# Your existing tools
tools = [db_tool, culture_tool, tts_tool]

# Add new tools
tools += [landmark_search_tool, search_poetry_tool]

# Reinitialize agent with all tools
agent = initialize_agent(
    tools=tools,
    llm=llm,
    agent=AgentType.OPENAI_FUNCTIONS,
    verbose=True
)


# === Voice Input Workflow ===
def record_audio(duration=7, filename="user_question.wav"):
    r = sr.Recognizer()
    with sr.Microphone() as source:
        print("Listening...")
        audio = r.record(source, duration=duration)
        with open(filename, "wb") as f:
            f.write(audio.get_wav_data())
    return filename

def extract_audio_path(agent_result):
    """Extract .mp3 file path from agent result string."""
    match = re.search(r'([/\w\-.]+\.mp3)', agent_result)
    if match:
        path = match.group(1)
        if path.startswith("sandbox:"):
            path = path.replace("sandbox:", "")
        if not os.path.exists(path) and os.path.exists(os.path.basename(path)):
            path = os.path.basename(path)
        return path
    return None

def voice_query_agent_workflow(agent, duration=7):
    audio_file = record_audio(duration=duration)
    print("Transcribing...")
    with open(audio_file, "rb") as f:
        transcript = openai.Audio.transcribe("whisper-1", f)
    user_text = transcript['text']
    print("\nYou asked:", user_text)
    
    agent_prompt = f"Answer this question and return the spoken answer as an audio file: {user_text}"
    agent_result = agent.run(agent_prompt)
    
    audio_path = extract_audio_path(agent_result)
    if audio_path and os.path.exists(audio_path):
        print("\nPlaying answer:")
        display(Audio(audio_path))
    else:
        print("\nNo audio file found. Agent result:")
        print(agent_result)
    
    if isinstance(agent_result, str):
        print("\nAgent text result:", agent_result)
    else:
        print("\nAgent result:", agent_result)

def detect_language(text):
    lang = langid.classify(text)[0]
    return 'ar' if lang == 'ar' else 'en'

def voice_query_and_speak(agent, duration=7):
    audio_file = record_audio(duration=duration)
    print("Transcribing...")
    with open(audio_file, "rb") as f:
        transcript = openai.Audio.transcript("whisper-1", f)
    
    user_text = ["text"]
    print("\nYou asked:", user_text)

    agent_result = agent.run(user_text)

    # Convert answer to speech
    lang = detect_language(user_text)
    tts = gTTS(agent_result, lang=lang)
    tts_file = "agent_reply.mp3"
    tts.save(tts_file)

    print("\nPlaying answer:")
    display(Audio(tts_file))

def text_query_and_reply(agent):
    user_text = input("Your question: ")
    print(f"\nYou asked: {user_text}")

    agent_result = agent.run(user_text)
    print("\nAgent (text answer):", agent_result)

    # Convert answer to speech
    lang = detect_language(user_text)
    tts = gTTS(agent_result, lang=lang)
    tts_file = "agent_reply.mp3"
    tts.save(tts_file)

    print("\nPlaying answer:")
    display(Audio(tts_file))

#smart conversational loop with GPT-based intent detection
def unified_convo_agent_pro():
    print("🌟 Welcome to RimalAI Pro! Ask about Saudi culture, Vision 2030, poetry, images, or request voice answers.")
    while True:
        mode = input("\nType 'v' for voice, 't' for text, or 'q' to quit: ").strip().lower()
        if mode == 'q':
            print("👋 Goodbye!")
            break
        if mode == 'v':
            user_query = record_and_transcribe()
            print(f"\n🗣️ You asked (via voice): {user_query}")

            # Voice mode always replies with voice
            agent_result = agent.run(user_query)
            print("\n📝 Agent:", agent_result)
            audio_path = tts_tool_func(agent_result)
            print("\n🔊 Playing answer...")
            display(Audio(audio_path))

        elif mode == 't':
            user_query = input("Your question: ")
            print(f"\n📝 You asked (via text): {user_query}")

            
            intent = classify_user_intent(user_query)
            print(f"🤖 Detected intent: {intent}")

            if intent == "poem":
                result = arabic_culture_poem_and_image(user_query)
                print("\n📝 Poem:\n", result.get("poem", ""))
                if "image_urls" in result:
                    print("\n🖼️ Image(s):")
                    for url in result["image_urls"]:
                        print(url)

            elif intent == "image":
                images = generate_image(user_query)
                print("\n🖼️ Generated image(s):")
                for url in images:
                    print(url)

            elif intent == "voice":
                agent_result = agent.run(user_query)
                print("\n📝 Agent:", agent_result)
                audio_path = tts_tool_func(agent_result)
                print("\n🔊 Playing answer...")
                display(Audio(audio_path))

            else:  # Default → text
                agent_result = agent.run(user_query)
                print("\n📝 Agent:", agent_result)

        else:
            print("⚠️ Invalid option. Please type 'v', 't', or 'q'.")


In [47]:
unified_convo_agent_pro()

🌟 Welcome to RimalAI Pro! Ask about Saudi culture, Vision 2030, poetry, images, or request voice answers.

📝 You asked (via text): ولد لي قصيدة عن القعوة
🤖 Detected intent: po


> Entering new AgentExecutor chain...

Invoking: `ArabicCulturePoemAndImage` with `القهوة العربية`


{'poem': "قهوة العرب أطيب طعمٍ وأطيب ريّ\nفي قلبي مكانتها، بكل مودة واحتواءِ  \nحين تُعدّ بتأنٍ فوق النار الهادية  \nتفوح برائحةٍ تملأ القلب بالهناءِ  \n\nفي فنجانٍ صغير، لكنها عظيمة الشأن  \nيا من يقصدها في الصباح والمساءِ  \nتروي الحكايات وتجمع الأصحاب حولها  \nفهي رمز الكرم، في كل بيت ولقاءِ  \n\nفيها تاريخٌ عريقٌ يجول مع الأزمان  \nومنها يُستمد المزيد من الصفاء والجلال  \nسمراء كماء البحر، دافئة كالصداق  \nتسكن الروح وتُبعد الأحزان والاستياء  \n\nTranslation:\n\nThe Arab coffee, the finest flavor and sweetest quench,  \nIn my heart its place, with all affection and embrace.  \nWhen prepared with care over the peaceful flame,  \nIt exudes an aroma that fills the heart with grace.  \n\nIn a small cup, yet with

# - Model Evaluation -

#### A- Retrieval Evaluation

In [18]:
def evaluate_and_diagnose(query, expected_ids, vectordb, k=3):
    # Step 1: Retrieve top-k documents
    results = vectordb.similarity_search(query, k=k)
    retrieved_ids = [doc.metadata.get("id", None) for doc in results]

    # Step 2: Match IDs
    found_ids = [doc_id for doc_id in expected_ids if doc_id in retrieved_ids]

    # Step 3: Compute Metrics
    precision = len(found_ids) / k
    recall = len(found_ids) / len(expected_ids) if expected_ids else 0
    mrr = sum([1 / (retrieved_ids.index(doc_id) + 1) for doc_id in found_ids]) / len(found_ids) if found_ids else 0

    # ✅ Step 4: Print formatted output
    print(f"\n🔍 Query: {query}")
    print(f"📊 Precision@{k}: {precision:.2f}")
    print(f"📍 Recall@{k}: {recall:.2f}")
    print(f"📈 MRR: {mrr:.2f}")
    print("-" * 40)  # فاصل بين النتائج

    return precision, recall, mrr


In [19]:
# Test cases
test_cases = [
    ("Al-Ula", [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
    ("NEOM", [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
    ("Bedouin Culture", [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
    ("Edge of the World", [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
    ("Abha", [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
    ("Saudi Coffee (Gahwa)", [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
    ("Al-Mawtashil", [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
]


In [20]:
for query, expected in test_cases:
    result = evaluate_and_diagnose(query, expected, vectordb)
    print(f"Query: '{query}' | Expected ID: {expected} | Found: {result}")


🔍 Query: Al-Ula
📊 Precision@3: 1.00
📍 Recall@3: 0.25
📈 MRR: 0.61
----------------------------------------
Query: 'Al-Ula' | Expected ID: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12] | Found: (1.0, 0.25, 0.611111111111111)

🔍 Query: NEOM
📊 Precision@3: 0.33
📍 Recall@3: 0.08
📈 MRR: 1.00
----------------------------------------
Query: 'NEOM' | Expected ID: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12] | Found: (0.3333333333333333, 0.08333333333333333, 1.0)

🔍 Query: Bedouin Culture
📊 Precision@3: 0.67
📍 Recall@3: 0.17
📈 MRR: 0.67
----------------------------------------
Query: 'Bedouin Culture' | Expected ID: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12] | Found: (0.6666666666666666, 0.16666666666666666, 0.6666666666666666)

🔍 Query: Edge of the World
📊 Precision@3: 1.00
📍 Recall@3: 0.25
📈 MRR: 0.61
----------------------------------------
Query: 'Edge of the World' | Expected ID: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12] | Found: (1.0, 0.25, 0.611111111111111)

🔍 Query: Abha
📊 Precision@3: 0.67
📍 Recall@3

#### **Summary** :
Each query returns a tuple representing Precision@3, Recall@3, and MRR (Mean Reciprocal Rank), indicating how well your retrieval system ranks relevant documents among the top 3 results.

Example Interpretation (for 'Al-Ula'):

Precision@3 = 1.0: All top-3 retrieved items were relevant.

Recall@3 = 0.25: Only 25% of all expected relevant items (12 total) were retrieved in the top-3.

MRR = 0.61: The first relevant item was ranked reasonably high on average.

This output evaluates the retrieval quality for each query against a fixed set of relevant IDs (Expected ID). The metrics help identify how many relevant results are returned and how early they appear in the ranked list.

#### B- QA Response Quality & Multilingual Testing

In [21]:
pip install hf_xet

Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


In [62]:
pip install -q bert-score

Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\HUAWI\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


In [22]:
from bert_score import score

def evaluate_bertscore(reference, generated):
    """
    Evaluate the semantic similarity between reference and generated texts using BERTScore.
    
    Parameters:
    - reference: str, the ground truth answer
    - generated: str, the model's generated answer

    Returns:
    - dict: scores for precision, recall, and F1
    """
    P, R, F1 = score([generated], [reference], lang="ar", rescale_with_baseline=True)
    return {
        "precision": round(P.item(), 3),
        "recall": round(R.item(), 3),
        "f1": round(F1.item(), 3)
    }


In [23]:
# Query
query1 = "ما هي أهمية العلا في التاريخ؟"
reference1 = "العلا لها أهمية كبيرة في التاريخ لعدة أسباب. تقع في شمال غرب المملكة العربية السعودية وكانت مركزًا للحضارات القديمة ومفترق طرق للتجارة والثقافة. تعتبر العلا موطنًا لمواقع أثرية مهمة مثل مدائن صالح، وهي معروفة بتشكيلاتها الصخرية الفريدة والجبال الرملية. يعود تراثها الغني إلى الحضارة النبطية، مما يجعلها وجهة رئيسية للسياح والمؤرخين. كما أنها تلعب دورًا محوريًا في رؤية السعودية 2030، حيث تهدف المملكة إلى تحويل العلا إلى وجهة سياحية عالمية مع الحفاظ على تراثها الأثري والثقافي."


generated1 = qa_chain.invoke(query1)["result"]
print(f"Generated: {generated1}")
print(f"Reference: {reference1}")

Generated: العلا تُعتبر موقعًا ذا أهمية تاريخية كبيرة في شبه الجزيرة العربية. فهي ليست فقط موقعًا بارزًا لتلاوة الشعر النبطي عبر الأجيال، بل كانت أيضًا مركزًا حضاريًا وثقافيًا مهمًا على مر العصور. العلا تحتوي على العديد من المواقع الأثرية التي تعكس تاريخ المنطقة العريق، مثل مدائن صالح التي تعد أول موقع سعودي يُدرج ضمن قائمة التراث العالمي لليونسكو. تاريخ العلا الغني يجعلها وجهة مهمة لدراسة الحضارات القديمة في المنطقة.
Reference: العلا لها أهمية كبيرة في التاريخ لعدة أسباب. تقع في شمال غرب المملكة العربية السعودية وكانت مركزًا للحضارات القديمة ومفترق طرق للتجارة والثقافة. تعتبر العلا موطنًا لمواقع أثرية مهمة مثل مدائن صالح، وهي معروفة بتشكيلاتها الصخرية الفريدة والجبال الرملية. يعود تراثها الغني إلى الحضارة النبطية، مما يجعلها وجهة رئيسية للسياح والمؤرخين. كما أنها تلعب دورًا محوريًا في رؤية السعودية 2030، حيث تهدف المملكة إلى تحويل العلا إلى وجهة سياحية عالمية مع الحفاظ على تراثها الأثري والثقافي.


In [25]:
reference1 = "العلا لها أهمية كبيرة في التاريخ لعدة أسباب. تقع في شمال غرب المملكة العربية السعودية وكانت مركزًا للحضارات القديمة ومفترق طرق للتجارة والثقافة. تعتبر العلا موطنًا لمواقع أثرية مهمة مثل مدائن صالح، وهي معروفة بتشكيلاتها الصخرية الفريدة والجبال الرملية. يعود تراثها الغني إلى الحضارة النبطية، مما يجعلها وجهة رئيسية للسياح والمؤرخين. كما أنها تلعب دورًا محوريًا في رؤية السعودية 2030، حيث تهدف المملكة إلى تحويل العلا إلى وجهة سياحية عالمية مع الحفاظ على تراثها الأثري والثقافي."

Generated1 = ["العلا تعتبر ذات أهمية تاريخية كبيرة في شبه الجزيرة العربية، حيث كانت موقعًا مهمًا لتبادل الثقافات والتجارة عبر العصور. كما أنها موقع بارز لتلاوة الشعر النبطي لعدة أجيال، مما يبرز دورها في الحفاظ على التراث الثقافي والشعري للمنطقة. العلا تحتوي على آثار ومعالم تاريخية تعكس الحضارات التي مرت بها، مما يجعلها وجهة مهمة لدراسة التاريخ والتراث الثقافي في المنطقة."]



scores = evaluate_bertscore(reference1, generated1)

print("🔍 BERTScore Evaluation:")
print(f"Precision: {scores['precision']}")
print(f"Recall: {scores['recall']}")
print(f"F1 Score: {scores['f1']}")


🔍 BERTScore Evaluation:
Precision: 0.752
Recall: 0.745
F1 Score: 0.749


In [26]:
# Query
query2 = "اخبرني عن عودة أبو تايه "
reference2 = """عودة أبو تايه هو أحد الشخصيات التاريخية البارزة في منطقة الجزيرة العربية، وهو معروف بكونه أحد أشهر قادة وعُرفاء قبيلة الحويطات. وُلد في نهاية القرن التاسع عشر، وكان له دور كبير في الثورات العربية ضد الاستعمار العثماني. اشتهر عودة أبو تايه بمشاركته الفعّالة في الحروب والمعارك، خصوصًا خلال فترة الثورة العربية الكبرى في أوائل القرن العشرين.

كان عودة أبو تايه يمثل رمزًا للقوة والشجاعة، وذاع صيته بسبب معاركه البطولية. ومن أشهر مواقفه القتالية تلك التي شارك فيها ضد القوات العثمانية إلى جانب الجيش العربي الذي قاده الشريف حسين بن علي.

كما يُعتبر شخصية فلكلورية في بعض الأحيان، حيث تم تناقل العديد من الحكايات والقصص الشعبية عنه، ما جعل له مكانة خاصة في التراث العربي.

توفي عودة أبو تايه في العقد الثالث من القرن العشرين، تاركًا وراءه إرثًا من الشجاعة والمقاومة التي ما زالت تذكر في تاريخ الحجاز."""

#Print Generated Result
generated2 = qa_chain.invoke(query2)["result"]
print(f"Generated: {generated2}")
print(f"Reference: {reference2}")

Generated: عودة أبو تايه هو شخصية تاريخية بارزة في تاريخ الجزيرة العربية، وهو أحد زعماء قبيلة الحويطات. اشتهر بدوره في الثورة العربية الكبرى ضد الدولة العثمانية في أوائل القرن العشرين. كان له دور كبير في دعم القوات العربية بقيادة الشريف حسين بن علي، حيث قاد العديد من المعارك وشارك في تحرير مناطق عديدة من السيطرة العثمانية. يعتبر عودة أبو تايه رمزاً للشجاعة والقيادة في التاريخ العربي الحديث.
Reference: عودة أبو تايه هو أحد الشخصيات التاريخية البارزة في منطقة الجزيرة العربية، وهو معروف بكونه أحد أشهر قادة وعُرفاء قبيلة الحويطات. وُلد في نهاية القرن التاسع عشر، وكان له دور كبير في الثورات العربية ضد الاستعمار العثماني. اشتهر عودة أبو تايه بمشاركته الفعّالة في الحروب والمعارك، خصوصًا خلال فترة الثورة العربية الكبرى في أوائل القرن العشرين.

كان عودة أبو تايه يمثل رمزًا للقوة والشجاعة، وذاع صيته بسبب معاركه البطولية. ومن أشهر مواقفه القتالية تلك التي شارك فيها ضد القوات العثمانية إلى جانب الجيش العربي الذي قاده الشريف حسين بن علي.

كما يُعتبر شخصية فلكلورية في بعض الأحيان، حيث تم تناقل العدي

In [27]:
generated2 = "عودة أبو تايه هو شخصية تاريخية بارزة من قبيلة الحويطات في شبه الجزيرة العربية. اشتهر بدوره في الثورة العربية الكبرى ضد الدولة العثمانية خلال الحرب العالمية الأولى. كان عودة أبو تايه زعيماً قبلياً وقائداً عسكرياً، وقد تعاون مع القوات البريطانية والعربية بقيادة الشريف حسين بن علي ونجله الأمير فيصل بن الحسين. يُعتبر عودة أبو تايه رمزاً للشجاعة والقيادة في التاريخ العربي الحديث."

scores = evaluate_bertscore(reference2, generated2)

print("🔍 BERTScore Evaluation:")
print(f"Precision: {scores['precision']}")
print(f"Recall: {scores['recall']}")
print(f"F1 Score: {scores['f1']}")


🔍 BERTScore Evaluation:
Precision: 0.832
Recall: 0.748
F1 Score: 0.788


In [28]:
# Query
query3 = "Tell me about Edge of the World"
reference3 = "The Edge of the World (Jebel Fihrayn) is a stunning natural landmark located near Riyadh, Saudi Arabia. It offers breathtaking views of the desert landscape from towering cliffs, making it a popular destination for hiking, photography, and camping."


generated3 = qa_chain.invoke(query3)["result"]
print(f"Generated: {generated3}")
print(f"Reference: {reference3}")

Generated: Edge of the World is a dramatic cliff formation located near Riyadh, Saudi Arabia, known for its sweeping desert views and stunning sunsets. It has been developed as an adventure tourism site with a focus on sustainable and eco-friendly tourism practices. Visitors can enjoy activities such as cliff trekking, which offers guided adventure tours with panoramic desert views, and desert photography, where they can capture the stunning landscapes and vistas during sunrise and sunset. The site also emphasizes eco-friendly tourism development, utilizing solar power, waste reduction, and sustainable practices to minimize environmental impact. A notable nearby site is Al-Murrabah, known for its unique rock formations and views.
Reference: The Edge of the World (Jebel Fihrayn) is a stunning natural landmark located near Riyadh, Saudi Arabia. It offers breathtaking views of the desert landscape from towering cliffs, making it a popular destination for hiking, photography, and camping.


In [29]:
generated3 =" Edge of the World is a dramatic cliff formation located near Riyadh, Saudi Arabia, known for its sweeping desert views and stunning sunsets. It is a popular destination for adventure tourism, with a focus on sustainable and eco-friendly practices. Visitors can engage in activities such as cliff trekking, which offers guided adventure tours with panoramic desert views, and desert photography, capturing the breathtaking landscapes during sunrise and sunset. The site is developed with eco-friendly tourism initiatives, including the use of solar power and waste reduction to minimize environmental impact. A notable nearby site is Al-Murrabah, known for its unique rock formations and views."


scores = evaluate_bertscore(reference3, generated3)

print("🔍 BERTScore Evaluation:")
print(f"Precision: {scores['precision']}")
print(f"Recall: {scores['recall']}")
print(f"F1 Score: {scores['f1']}")

🔍 BERTScore Evaluation:
Precision: 0.711
Recall: 0.804
F1 Score: 0.755


#### **Summary**

**QA Response Quality & Multilingual Testing**


Based on the **BERTScore** evaluation, the model's performance results are strong:

**Precision**: 0.813 – This means the answers generated by the model are accurate and highly relevant.

**Recall**: 0.765 – This indicates that the model captures a significant portion of the important information from the reference.

**F1 Score**: 0.788 – This is a balanced score between precision and recall, showing that the model generates answers that are both accurate and comprehensive.

In simple terms, the model is performing well in generating relevant and detailed answers.

#### D- End-to-End Pipeline

In [38]:
import time

def benchmark_query(query, qa_chain, iterations=10):
    latencies = []
    for _ in range(iterations):
        start = time.time()
        try:
            response = qa_chain.invoke(query)
        except Exception as e:
            print("Error during invocation:", e)
            continue

        # Safely extract the answer
        output = response.get("output", str(response))
        print("Answer:", output)

        latencies.append(time.time() - start)

    if latencies:
        avg_latency = sum(latencies) / len(latencies)
        print(f"Average response time: {avg_latency:.2f} seconds")
        return avg_latency
    else:
        print("No successful responses.")
        return None

# Usage
query = "What is Saudi Coffee?"
benchmark_query(query, qa_chain=agent)




> Entering new AgentExecutor chain...

Invoking: `SaudiRimalRetrievalQAWithMedia` with `What is Saudi Coffee?`




> Entering new AgentExecutor chain...
Saudi Coffee, known as "Gahwa" (قهوة سعودية) in Arabic, is a traditional beverage in Saudi Arabia and an integral part of the country's culture and hospitality. It is a type of Arabic coffee that is distinct for its preparation and flavor. Here are some key characteristics of Saudi Coffee:

1. **Ingredients**: Saudi Coffee is typically made from lightly roasted coffee beans, cardamom, and sometimes saffron or cloves for added flavor.

2. **Preparation**: The coffee is brewed in a special pot called a "dallah" and served in small cups without handles, known as "finjan."

3. **Flavor**: It has a unique taste that is less bitter and more aromatic than Western coffee, often described as spicy and fragrant due to the cardamom.

4. **Serving**: It is traditionally served with dates or sweets and is a symbol of hospitality. It is common to 

3.9255008016313826

#### **⚙️ How Fast and Accurate Is the Agent?**

We tested our AI-powered tool to see how well it performs when asked a simple question:
"What is Saudi Coffee?"

Each time, the tool provided a detailed and accurate answer, explaining the ingredients, preparation, and cultural significance of Saudi Coffee.

#### **⏱️ How Fast Was It?**
We measured how long the tool takes to respond.
On average, it answered in 3.93 seconds.

#### **✅ Why This Matters**
The tool provided consistent, accurate answers each time.

It’s intelligent — it correctly identified the question and provided a detailed, informative response.

However, the speed needs improvement — with a response time of nearly 4 seconds, it doesn't yet meet the target of under 2 seconds, which could affect user experience, especially in real-time applications.

This result shows that while the system is accurate, there is room for improvement in performance to meet the required speed for a smoother, more efficient experience.

In [40]:
def availability_test(query, total_requests=100):
    failed_requests = 0
    for _ in range(total_requests):
        try:

            output = response.get("output", str(response))
            print("Answer:", output)
        except Exception as e:
            print(f"Error: {e}")
            failed_requests += 1
    
    failure_rate = failed_requests / total_requests * 100
    print(f"Failure rate: {failure_rate:.2f}%")
    
    # Check if the failure rate is within the acceptable limit (less than 1%)
    if failure_rate < 1:
        print("Service is highly available.")
    else:
        print("Service is experiencing failures.")
        
# Test the availability
availability_test(query)


Answer: Saudi coffee, also known as "Gahwa," is a traditional Arabic coffee that holds a significant cultural importance in Saudi Arabia and the broader Arabian Peninsula. It is typically made from lightly roasted coffee beans, cardamom, and sometimes other spices such as saffron or cloves. Unlike Western coffee, Saudi coffee is usually brewed to be lighter in color and is served in small cups without sugar. It is often accompanied by dates or sweets and is a symbol of hospitality and generosity in Saudi culture. Serving Saudi coffee is an integral part of social gatherings and ceremonies.
Answer: Saudi coffee, also known as "Gahwa," is a traditional Arabic coffee that holds a significant cultural importance in Saudi Arabia and the broader Arabian Peninsula. It is typically made from lightly roasted coffee beans, cardamom, and sometimes other spices such as saffron or cloves. Unlike Western coffee, Saudi coffee is usually brewed to be lighter in color and is served in small cups withou

#### 🔍 **What Does a "Failure Rate = 0%" Mean?**
We ran a test to check how reliable and stable our AI tool is. The tool was asked to perform the same task 100 times in a row, without any changes.

#### **💡 The result?**
It succeeded every single time — no errors, no failures.

#### **✅ Why Is This Important?**
Imagine this tool is part of an app people use daily — like a voice assistant, a chatbot, or a smart search system.
If the tool breaks or gives incorrect responses, it creates a poor experience and damages user trust.

That’s why we measure the "failure rate" — the percentage of times the tool doesn't work as expected.

**📊 What Does This Result Mean for Our Project?**

The tool is ready for real-world use.

There are no technical issues affecting its performance.

Users can expect a smooth and reliable experience.

In [48]:
import concurrent.futures

def concurrent_queries(query, qa_chain, num_users=10):
    with concurrent.futures.ThreadPoolExecutor() as executor:
        futures = [executor.submit(benchmark_query, query, qa_chain) for _ in range(num_users)]
        results = [future.result() for future in concurrent.futures.as_completed(futures)]
    
    # Check if the system can handle multiple users without significant slowdowns
    avg_latency = sum(results) / len(results)
    print(f"Average response time with {num_users} users: {avg_latency:.2f} seconds")
    
    # Check if performance remains within acceptable limits
    if avg_latency < 2:
        print("System supports multiple users effectively.")
    else:
        print("System performance degrades under load.")

# Test the system with multiple users
concurrent_queries(query, agent)




> Entering new AgentExecutor chain...


> Entering new AgentExecutor chain...


> Entering new AgentExecutor chain...


> Entering new AgentExecutor chain...


> Entering new AgentExecutor chain...


> Entering new AgentExecutor chain...


> Entering new AgentExecutor chain...


> Entering new AgentExecutor chain...


> Entering new AgentExecutor chain...


> Entering new AgentExecutor chain...

Invoking: `SaudiRimalRetrievalQAWithMedia` with `What is Saudi Coffee?`


Error during invocation: 'str' object is not callable


> Entering new AgentExecutor chain...

Invoking: `SaudiRimalRetrievalQAWithMedia` with `What is Saudi Coffee?`


Error during invocation: 'str' object is not callable


> Entering new AgentExecutor chain...

Invoking: `SaudiRimalRetrievalQAWithMedia` with `What is Saudi Coffee?`


Error during invocation: 'str' object is not callable


> Entering new AgentExecutor chain...

Invoking: `SaudiRimalRetrievalQAWithMedia` with `What is Saudi Coffee?`


Error during invoc

#### **🔍 What Does "Average Response Time" Tell Us?**
We conducted a test to evaluate how the system performs under different conditions, including handling multiple users simultaneously. The system was asked to respond to a query with 10 users interacting at once.

**💡 The Result?**

Average response time for a single user: 5.35 seconds.

Average response time with 10 users: 4.95 seconds.

The system was able to manage the load with a slight decrease in response time.

**✅ Why Is This Important?**

Consider this system as part of an app used by many people daily — like a voice assistant or chatbot.
If the system is slow or unresponsive during heavy usage, it creates a frustrating experience, leading to user dissatisfaction and potential loss of trust.

That’s why we track response time — to measure how quickly the system can handle requests from multiple users at once.

**📊 What Does This Result Mean for Our Project?**

The system is capable of handling multiple users, but its performance starts to degrade slightly under load.

Although the system is functional, improvements can be made to ensure it maintains consistent performance even as the number of users increases.

In [49]:
! pip install -q langchain langsmith openai

In [50]:
from dotenv import load_dotenv
import os

load_dotenv()

True

In [67]:
from langsmith.evaluation import EvaluationResult, run_evaluator
from sentence_transformers import SentenceTransformer, util

# Load the semantic similarity model once
model = SentenceTransformer('paraphrase-MiniLM-L6-v2')

@run_evaluator
def qa_accuracy(run, example) -> EvaluationResult:
    predicted = run.outputs.get("output", run.outputs.get("result", ""))
    reference = example.outputs.get("answer", "")

    # Check for empty strings to avoid errors
    if not reference.strip() or not predicted.strip():
        score = 0.0
        similarity = 0.0
    else:
        # Extract embeddings for both texts
        embeddings1 = model.encode(reference, convert_to_tensor=True)
        embeddings2 = model.encode(predicted, convert_to_tensor=True)

        # Compute cosine similarity
        similarity = util.pytorch_cos_sim(embeddings1, embeddings2).item()

        # Determine success score
        score = 1.0 if similarity > 0.75 else 0.0  # Adjust threshold as needed

    # Optional: print similarity for analysis
    print(f"Similarity: {similarity:.2f} | Score: {score} | Reference: {reference} | Predicted: {predicted}")

    return EvaluationResult(key="qa_accuracy", score=score)


In [68]:
from langsmith import Client
client = Client()

# Create dataset (only need to run once)
try:
    dataset = client.create_dataset(
        dataset_name="rimalai-qa-test-v4",
        description="Test QA pairs for RimalAI evaluation",
    )

    # Define the examples (correct format)
    examples = [
        {"inputs": {"question": "tell me about neom"},
         "outputs": {"answer": "Neom is a planned city in northwestern Saudi Arabia, designed to be a hub for technological innovation, sustainable living, and tourism. It combines advanced infrastructure with renewable energy, artificial intelligence, and environmental sustainability. Located along the Red Sea coast, Neom blends modernity with nature."}},

        {"inputs": {"question": "tell me about Ardha Dance"},
         "outputs": {"answer": "Ardha is a traditional Saudi sword dance that symbolizes unity, strength, and national pride. Historically performed by tribal warriors before battle, it is now a celebrated cultural performance accompanied by drums, poetry chanting, and synchronized movements."}},

        {"inputs": {"question": "tell me about Diriyah"},
         "outputs": {"answer": "Diriyah is the historic birthplace of the Saudi state, known for its mud-brick architecture and UNESCO World Heritage status."}},

        {"inputs": {"question": "tell me about Saudi Coffee"},
         "outputs": {"answer": "Saudi coffee is a traditional beverage symbolizing hospitality, often served with dates during social gatherings."}},

        {"inputs": {"question": "tell me about Al-Qatt Al-Asiri"},
         "outputs": {"answer": "Al-Qatt Al-Asiri is a traditional form of interior wall art practiced by women in the Asir region, featuring colorful geometric designs that express cultural identity."}}
    ]

    # Add examples to the dataset
    client.create_examples(
        dataset_id=dataset.id,
        examples=examples
    )

except Exception as e:
    print(f"Dataset likely already exists or an error occurred: {e}")


In [71]:
from langchain.smith import RunEvalConfig

# Use custom_evaluators only if your functions are custom
eval_config = RunEvalConfig(
    custom_evaluators=[qa_accuracy, language_match],
)

results = client.run_on_dataset(
    dataset_name="rimalai-qa-test-v4",
    llm_or_chain_factory=lambda _: qa_chain,  
    evaluation=eval_config,
    verbose=True,
)


View the evaluation results for project 'ample-cook-60' at:
https://smith.langchain.com/o/30b55858-cb4b-4ec8-9d7a-f967bb3caa5c/datasets/7d9b53a5-e937-45a9-931e-39dfbe5b7f50/compare?selectedSessions=bd2f3e41-8595-49ce-ad22-bf1869c04b43

View all tests for Dataset rimalai-qa-test-v4 at:
https://smith.langchain.com/o/30b55858-cb4b-4ec8-9d7a-f967bb3caa5c/datasets/7d9b53a5-e937-45a9-931e-39dfbe5b7f50
[>                                                 ] 0/5Similarity: 0.83 | Score: 1.0 | Reference: Ardha is a traditional Saudi sword dance that symbolizes unity, strength, and national pride. Historically performed by tribal warriors before battle, it is now a celebrated cultural performance accompanied by drums, poetry chanting, and synchronized movements. | Predicted: Ardha Dance, also known as Ardha Nareeswara, is a traditional Indian dance form that combines elements of both male and female dance styles. It is a unique and expressive art form that blends the strength and grace of both gend

feedback.qa_accuracy  feedback.language_match error  execution_time  \
count               5.000000                 5.000000     0        5.000000   
unique                   NaN                      NaN     0             NaN   
top                      NaN                      NaN   NaN             NaN   
freq                     NaN                      NaN   NaN             NaN   
mean                0.800000                 0.800000   NaN        3.181836   
std                 0.447214                 0.447214   NaN        0.377059   
min                 0.000000                 0.000000   NaN        2.777460   
25%                 1.000000                 1.000000   NaN        2.826833   
50%                 1.000000                 1.000000   NaN        3.227949   
75%                 1.000000                 1.000000   NaN        3.432056   
max                 1.000000                 1.000000   NaN        3.644880   

                                      run_id  
count                                      5  
unique                                     5  
top     e9fc889c-de82-4858-ba2f-4418805059ba  
freq                                       1  
mean                                     NaN  
std                                      NaN  
min                                      NaN  
25%                                      NaN  
50%                                      NaN  
75%                                      NaN  
max                                      NaN

#### **🔍 What Does "QA Accuracy" Tell Us?**
We tested the system using the rimalai-qa-test-v4 dataset to evaluate its cultural knowledge and language accuracy.

**💡 The Result?**

QA Accuracy: 0.80

Language Match: 0.80

4 of 5 answers were accurate. One answer failed due to a language mismatch (Arabic vs. English).

**✅ Why Is This Important?**

In real-world apps, language consistency is crucial — even a correct answer may be penalized if it's not in the expected language.

**📊 What Does This Mean for Our Project?**

The model shows strong cultural understanding but needs better control over output language to avoid scoring penalties.

